{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.0.6'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Project = '/Users/daehwa/Library/CloudStorage/OneDrive-Personal/Junlab/Projects/Adipocyte'\n",
    "from IPython.display import display\n",
    "import pandas as pd\n",
    "import gseapy as gp\n",
    "import numpy as np\n",
    "import pickle\n",
    "\n",
    "gp.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ct/z73cq4_j2g5612tltg_pmxxc0000gn/T/ipykernel_99135/1937983951.py:28: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  MitoPW['gene_ids'][i] += name2ID[gene].replace('|',',') + ','\n",
      "/var/folders/ct/z73cq4_j2g5612tltg_pmxxc0000gn/T/ipykernel_99135/1937983951.py:29: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  MitoPW['gene_ids'][i] = MitoPW['gene_ids'][i][:-1]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Basal_L2FC</th>\n",
       "      <th>Puromycin_L2FC</th>\n",
       "      <th>Cycloheximide_L2FC</th>\n",
       "      <th>Mito_TargetP_score</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>RD4</th>\n",
       "      <th>RD8</th>\n",
       "      <th>Post-init</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000004610</td>\n",
       "      <td>0.172060</td>\n",
       "      <td>0.394754</td>\n",
       "      <td>0.501723</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Etfb</td>\n",
       "      <td>-0.510055</td>\n",
       "      <td>-0.754061</td>\n",
       "      <td>0.106969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000052738</td>\n",
       "      <td>0.397798</td>\n",
       "      <td>0.816989</td>\n",
       "      <td>0.362190</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Suclg1</td>\n",
       "      <td>-0.142135</td>\n",
       "      <td>-0.117711</td>\n",
       "      <td>-0.454799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000028959</td>\n",
       "      <td>-0.113037</td>\n",
       "      <td>0.236100</td>\n",
       "      <td>0.513938</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Fastk</td>\n",
       "      <td>-0.099066</td>\n",
       "      <td>-0.266143</td>\n",
       "      <td>0.277838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000024521</td>\n",
       "      <td>-0.006939</td>\n",
       "      <td>-0.008913</td>\n",
       "      <td>-0.055495</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pmaip1</td>\n",
       "      <td>0.275798</td>\n",
       "      <td>0.504255</td>\n",
       "      <td>-0.046581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000026618</td>\n",
       "      <td>3.803577</td>\n",
       "      <td>0.081617</td>\n",
       "      <td>5.188976</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Iars2</td>\n",
       "      <td>0.029945</td>\n",
       "      <td>0.292534</td>\n",
       "      <td>5.107359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778</th>\n",
       "      <td>ENSMUSG00000030706</td>\n",
       "      <td>0.191132</td>\n",
       "      <td>0.580547</td>\n",
       "      <td>-0.183006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Mrpl48</td>\n",
       "      <td>-0.227354</td>\n",
       "      <td>-0.165974</td>\n",
       "      <td>-0.763553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>ENSMUSG00000030802</td>\n",
       "      <td>-0.006702</td>\n",
       "      <td>0.145352</td>\n",
       "      <td>2.307026</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Bckdk</td>\n",
       "      <td>0.076094</td>\n",
       "      <td>0.420585</td>\n",
       "      <td>2.161675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>ENSMUSG00000020268</td>\n",
       "      <td>-0.194344</td>\n",
       "      <td>-0.246781</td>\n",
       "      <td>-0.455046</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Lyrm7</td>\n",
       "      <td>0.086306</td>\n",
       "      <td>0.056225</td>\n",
       "      <td>-0.208265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>ENSMUSG00000024158</td>\n",
       "      <td>0.481418</td>\n",
       "      <td>0.749426</td>\n",
       "      <td>0.909822</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Hagh</td>\n",
       "      <td>-0.383863</td>\n",
       "      <td>-0.394963</td>\n",
       "      <td>0.160396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>ENSMUSG00000025465</td>\n",
       "      <td>-0.353842</td>\n",
       "      <td>0.548586</td>\n",
       "      <td>0.188394</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Echs1</td>\n",
       "      <td>0.176456</td>\n",
       "      <td>0.457182</td>\n",
       "      <td>-0.360191</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>783 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id  Basal_L2FC  Puromycin_L2FC  Cycloheximide_L2FC  \\\n",
       "0    ENSMUSG00000004610    0.172060        0.394754            0.501723   \n",
       "1    ENSMUSG00000052738    0.397798        0.816989            0.362190   \n",
       "2    ENSMUSG00000028959   -0.113037        0.236100            0.513938   \n",
       "3    ENSMUSG00000024521   -0.006939       -0.008913           -0.055495   \n",
       "4    ENSMUSG00000026618    3.803577        0.081617            5.188976   \n",
       "..                  ...         ...             ...                 ...   \n",
       "778  ENSMUSG00000030706    0.191132        0.580547           -0.183006   \n",
       "779  ENSMUSG00000030802   -0.006702        0.145352            2.307026   \n",
       "780  ENSMUSG00000020268   -0.194344       -0.246781           -0.455046   \n",
       "781  ENSMUSG00000024158    0.481418        0.749426            0.909822   \n",
       "782  ENSMUSG00000025465   -0.353842        0.548586            0.188394   \n",
       "\n",
       "     Mito_TargetP_score gene_name       RD4       RD8  Post-init  \n",
       "0                   1.0      Etfb -0.510055 -0.754061   0.106969  \n",
       "1                   1.0    Suclg1 -0.142135 -0.117711  -0.454799  \n",
       "2                   0.0     Fastk -0.099066 -0.266143   0.277838  \n",
       "3                   0.0    Pmaip1  0.275798  0.504255  -0.046581  \n",
       "4                   5.0     Iars2  0.029945  0.292534   5.107359  \n",
       "..                  ...       ...       ...       ...        ...  \n",
       "778                 0.0    Mrpl48 -0.227354 -0.165974  -0.763553  \n",
       "779                 4.0     Bckdk  0.076094  0.420585   2.161675  \n",
       "780                 4.0     Lyrm7  0.086306  0.056225  -0.208265  \n",
       "781                 5.0      Hagh -0.383863 -0.394963   0.160396  \n",
       "782                 5.0     Echs1  0.176456  0.457182  -0.360191  \n",
       "\n",
       "[783 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "##\n",
    "RD = pd.read_csv(f'{Project}/Results/Tables/CPM_RD_calculation/v20221217/adi_RD_avg_log2FC.tsv', sep='\\t').rename(columns={'D4':'RD4', 'D8':'RD8'})\n",
    "\n",
    "logFC = {}\n",
    "logFC['RD'] = RD\n",
    "\n",
    "##\n",
    "APEX = pd.read_excel(f'{Project}/Results/Other_papers/Fazal_Han_etal.,2019,cell/Tables/Resources/(Sup. table4) Atlas of subcellular RNA localization revealed by APEX-seq_Cell_2019.xlsx', sheet_name='OMM APEX-seq data')\n",
    "CHIP = pd.read_csv('/Users/daehwa/Library/CloudStorage/OneDrive-Personal/Junlab/Data_Library/GSEA/chip/Mouse_Ensembl_Gene_ID_Human_Orthologs_MSigDB.v2023.2.Hs.chip', sep='\\t', usecols=['Probe Set ID','Gene Symbol']).rename(columns={'Probe Set ID':'gene_id', 'Gene Symbol':'Common_Gene'})\n",
    "\n",
    "APEX2 = pd.merge(CHIP, APEX[['Common_Gene','Basal_L2FC','Puromycin_L2FC','Cycloheximide_L2FC','Mito_TargetP_score']], on='Common_Gene').sort_values('gene_id').drop_duplicates(subset='gene_id')\n",
    "\n",
    "APEX3 = pd.merge(APEX2[['gene_id','Basal_L2FC','Puromycin_L2FC','Cycloheximide_L2FC','Mito_TargetP_score']], logFC['RD'], on='gene_id')\n",
    "\n",
    "# filt mtDNA genes\n",
    "MitoCarta3_file = '/Users/daehwa/Library/CloudStorage/OneDrive-Personal/Junlab/Data_Library/MITOCARTA3.0/Mouse.MitoCarta3.0.xls'\n",
    "MitoCarta = pd.read_excel(MitoCarta3_file, 'A Mouse MitoCarta3.0')\n",
    "MitoPW    = pd.read_excel(MitoCarta3_file, 'C MitoPathways')\n",
    "\n",
    "name2ID = MitoCarta[['Symbol','EnsemblGeneID']].set_index('Symbol')['EnsemblGeneID'].to_dict()\n",
    "mito_genes = MitoCarta[['EnsemblGeneID']].drop_duplicates().rename(columns={'EnsemblGeneID':'gene_id'})\n",
    "\n",
    "MitoPW['PW level'] = MitoPW['MitoPathway Hierarchy'].str.split('>').str.len()\n",
    "MitoPW['gene_ids'] = ''\n",
    "for i in range(len(MitoPW)):\n",
    "    for gene in MitoPW.iloc[i,:]['Genes'].split(', '):\n",
    "        if type(name2ID[gene]) == float : continue\n",
    "        MitoPW['gene_ids'][i] += name2ID[gene].replace('|',',') + ','\n",
    "    MitoPW['gene_ids'][i] = MitoPW['gene_ids'][i][:-1]\n",
    "MitoPW = MitoPW.set_index('MitoPathway',drop=True)\n",
    "\n",
    "mtDNA = MitoCarta[['Symbol','EnsemblGeneID']]\n",
    "mtDNA = mtDNA[mtDNA['Symbol'].str.startswith('mt-')].reset_index(drop=True)\n",
    "mtDNA = mtDNA.rename(columns={'Symbol':'gene_name', 'EnsemblGeneID':'gene_id'})[['gene_id']]\n",
    "\n",
    "tmp = pd.merge(mito_genes, mtDNA, on='gene_id', how='outer', indicator=True).sort_values('_merge')\n",
    "mito_genes_nc = tmp[tmp['_merge']=='left_only'].drop(columns='_merge')[['gene_id']]\n",
    "\n",
    "tmp = pd.merge(APEX3, mtDNA, on='gene_id', how='outer', indicator=True).sort_values('_merge')\n",
    "APEX3 = tmp[tmp['_merge']=='left_only'].drop(columns='_merge')\n",
    "\n",
    "APEX3_cyto = pd.merge(APEX3, mito_genes_nc, on='gene_id', how='left', indicator=True)\n",
    "APEX3_cyto = APEX3_cyto[APEX3_cyto['_merge']=='left_only'].drop(columns='_merge')\n",
    "\n",
    "APEX3_mito = pd.merge(APEX3, mito_genes_nc, on='gene_id')\n",
    "\n",
    "# display(APEX3, APEX3_cyto, APEX3_mito)\n",
    "APEX3 = APEX3.sample(frac=1).reset_index(drop=True)\n",
    "APEX3_cyto = APEX3_cyto.sample(frac=1).reset_index(drop=True)\n",
    "APEX3_mito = APEX3_mito.sample(frac=1).reset_index(drop=True)\n",
    "\n",
    "APEX3_mito['Post-init'] = APEX3_mito['Cycloheximide_L2FC']-APEX3_mito['Puromycin_L2FC']\n",
    "\n",
    "display(APEX3_mito)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'APEXseq_Q4':                 gene_id\n",
       " 1    ENSMUSG00000052738\n",
       " 6    ENSMUSG00000030335\n",
       " 8    ENSMUSG00000051671\n",
       " 9    ENSMUSG00000038412\n",
       " 14   ENSMUSG00000044881\n",
       " ..                  ...\n",
       " 775  ENSMUSG00000030614\n",
       " 777  ENSMUSG00000070283\n",
       " 778  ENSMUSG00000030706\n",
       " 781  ENSMUSG00000024158\n",
       " 782  ENSMUSG00000025465\n",
       " \n",
       " [284 rows x 1 columns]}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genelists = {\n",
    "            #  'Q1_Q2' : APEX3_mito[APEX3_mito['Post-init']>1.5][['gene_id']],\n",
    "            #  'Q2' : APEX3_mito[(APEX3_mito['Post-init']>1.5) & (APEX3_mito['Puromycin_L2FC']<0.4)][['gene_id']],\n",
    "            #  'Q2_Q3' : APEX3_mito[APEX3_mito['Puromycin_L2FC']<0.4][['gene_id']],\n",
    "             'APEXseq_Q4' : APEX3_mito[(APEX3_mito['Post-init']<1.5) & (APEX3_mito['Puromycin_L2FC']>0.4)][['gene_id']],\n",
    "             }\n",
    "\n",
    "display(genelists)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Etfb',\n",
       " 'Suclg1',\n",
       " 'Fastk',\n",
       " 'Pmaip1',\n",
       " 'Iars2',\n",
       " 'Ccdc127',\n",
       " 'Mrpl51',\n",
       " 'Mrpl28',\n",
       " 'Coa6',\n",
       " 'Higd1a']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "783\n"
     ]
    }
   ],
   "source": [
    "conv = pd.read_csv('/Users/daehwa/Library/CloudStorage/OneDrive-Personal/Junlab/Data_Library/GSEA/chip/Mouse_Ensembl_Gene_ID_MSigDB.v2023.2.Mm.chip', sep='\\t')\n",
    "conv = conv.drop(columns='Gene Title').rename(columns={'Probe Set ID':'gene_id', 'Gene Symbol':'external_gene_name'})\n",
    "# display(conv)\n",
    "\n",
    "BG_genes = APEX3_mito[['gene_id']]\n",
    "BG_Symbols = pd.merge(BG_genes, conv, on='gene_id')\n",
    "BG_Symbols = BG_Symbols.to_dict('list')['external_gene_name']\n",
    "display(BG_Symbols[:10])\n",
    "print(len(BG_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "APEXseq_Q4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>OXPHOS subunits</td>\n",
       "      <td>46/60</td>\n",
       "      <td>3.628878e-11</td>\n",
       "      <td>3.193413e-09</td>\n",
       "      <td>2.331566</td>\n",
       "      <td>56.049712</td>\n",
       "      <td>Ndufb10;Ndufb7;Ndufa13;Cox6c;Sdhb;Ndufa6;Ndufs...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Mitochondrial ribosome</td>\n",
       "      <td>53/76</td>\n",
       "      <td>4.433454e-10</td>\n",
       "      <td>1.950720e-08</td>\n",
       "      <td>2.137311</td>\n",
       "      <td>46.030574</td>\n",
       "      <td>Mrpl34;Mrpl21;Mrps6;Mrpl13;Mrps18b;Gadd45gip1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>CI subunits</td>\n",
       "      <td>26/34</td>\n",
       "      <td>1.235894e-06</td>\n",
       "      <td>3.625288e-05</td>\n",
       "      <td>2.227091</td>\n",
       "      <td>30.296708</td>\n",
       "      <td>Ndufb10;Ndufb7;Ndufa13;Ndufa6;Ndufs5;Ndufb11;N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>CIV subunits</td>\n",
       "      <td>12/12</td>\n",
       "      <td>4.458087e-06</td>\n",
       "      <td>9.807792e-05</td>\n",
       "      <td>2.831193</td>\n",
       "      <td>34.882532</td>\n",
       "      <td>Cox7b;Cox8a;Cox5a;Cox7a2;Cox6c;Cox4i1;Cox6b1;C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>OXPHOS assembly factors</td>\n",
       "      <td>24/49</td>\n",
       "      <td>4.105660e-02</td>\n",
       "      <td>6.951296e-01</td>\n",
       "      <td>1.395549</td>\n",
       "      <td>4.455712</td>\n",
       "      <td>Tmem70;Uqcc2;Ndufaf4;Cmc1;Timmdc1;Coa4;Tmem126...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Respirasome assembly</td>\n",
       "      <td>3/3</td>\n",
       "      <td>4.739520e-02</td>\n",
       "      <td>6.951296e-01</td>\n",
       "      <td>2.772647</td>\n",
       "      <td>8.454449</td>\n",
       "      <td>Higd2a;Higd1a;Cox7a2l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>CIII subunits</td>\n",
       "      <td>6/9</td>\n",
       "      <td>6.230448e-02</td>\n",
       "      <td>7.832563e-01</td>\n",
       "      <td>1.902769</td>\n",
       "      <td>5.281557</td>\n",
       "      <td>Uqcr11;Cyc1;Uqcrfs1;Uqcrh;Uqcrq;Uqcr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>ROS and glutathione metabolism</td>\n",
       "      <td>8/23</td>\n",
       "      <td>6.378397e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>0.994844</td>\n",
       "      <td>0.447350</td>\n",
       "      <td>Hagh;Prdx4;Gpx4;Prdx5;Sod1;Prdx3;Gpx1;Msrb2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Q-linked reactions, other</td>\n",
       "      <td>1/5</td>\n",
       "      <td>8.956430e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>0.748918</td>\n",
       "      <td>0.082541</td>\n",
       "      <td>Etfa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Pyruvate metabolism</td>\n",
       "      <td>3/12</td>\n",
       "      <td>8.708325e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>0.767389</td>\n",
       "      <td>0.106134</td>\n",
       "      <td>Mpc1;Pdhb;Mpc2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Proteases</td>\n",
       "      <td>4/25</td>\n",
       "      <td>9.937830e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>0.477194</td>\n",
       "      <td>0.002976</td>\n",
       "      <td>Clpp;Lap3;Htra2;Parl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Polycistronic mtRNA processing</td>\n",
       "      <td>1/12</td>\n",
       "      <td>9.957247e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>0.326561</td>\n",
       "      <td>0.001399</td>\n",
       "      <td>Hsd17b10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Phospholipid metabolism</td>\n",
       "      <td>4/19</td>\n",
       "      <td>9.548283e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>0.628959</td>\n",
       "      <td>0.029073</td>\n",
       "      <td>Triap1;Chpt1;Stard7;Prelid1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Amidoxime reducing complex</td>\n",
       "      <td>1/2</td>\n",
       "      <td>5.941539e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>1.653968</td>\n",
       "      <td>0.861084</td>\n",
       "      <td>Cyb5b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>SAM</td>\n",
       "      <td>2/3</td>\n",
       "      <td>2.989931e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>1.973451</td>\n",
       "      <td>2.382616</td>\n",
       "      <td>Mtx1;Mtx2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>OXA</td>\n",
       "      <td>1/1</td>\n",
       "      <td>3.627075e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>2.760141</td>\n",
       "      <td>2.799220</td>\n",
       "      <td>Oxa1l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Nucleotide synthesis and processing</td>\n",
       "      <td>7/17</td>\n",
       "      <td>4.239890e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>1.183784</td>\n",
       "      <td>1.015743</td>\n",
       "      <td>Nt5dc2;Guk1;Nudt2;Dut;Suclg1;Hint1;Dguok</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Nucleotide import</td>\n",
       "      <td>1/7</td>\n",
       "      <td>9.579569e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>0.547795</td>\n",
       "      <td>0.023529</td>\n",
       "      <td>Slc25a5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Mitophagy</td>\n",
       "      <td>4/11</td>\n",
       "      <td>6.102665e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>1.077656</td>\n",
       "      <td>0.532211</td>\n",
       "      <td>Park7;Parl;Tomm7;Fundc1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>MITOCARTA3_MTPWnoChild_genename.gmt</td>\n",
       "      <td>Mitochondrial ribosome assembly</td>\n",
       "      <td>4/19</td>\n",
       "      <td>9.548283e-01</td>\n",
       "      <td>9.968845e-01</td>\n",
       "      <td>0.628959</td>\n",
       "      <td>0.029073</td>\n",
       "      <td>Mtg1;Malsu1;Tfb1m;Mpv17l2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Gene_set                                 Term  \\\n",
       "55  MITOCARTA3_MTPWnoChild_genename.gmt                      OXPHOS subunits   \n",
       "48  MITOCARTA3_MTPWnoChild_genename.gmt               Mitochondrial ribosome   \n",
       "5   MITOCARTA3_MTPWnoChild_genename.gmt                          CI subunits   \n",
       "10  MITOCARTA3_MTPWnoChild_genename.gmt                         CIV subunits   \n",
       "54  MITOCARTA3_MTPWnoChild_genename.gmt              OXPHOS assembly factors   \n",
       "63  MITOCARTA3_MTPWnoChild_genename.gmt                 Respirasome assembly   \n",
       "8   MITOCARTA3_MTPWnoChild_genename.gmt                        CIII subunits   \n",
       "62  MITOCARTA3_MTPWnoChild_genename.gmt       ROS and glutathione metabolism   \n",
       "61  MITOCARTA3_MTPWnoChild_genename.gmt            Q-linked reactions, other   \n",
       "60  MITOCARTA3_MTPWnoChild_genename.gmt                  Pyruvate metabolism   \n",
       "59  MITOCARTA3_MTPWnoChild_genename.gmt                            Proteases   \n",
       "58  MITOCARTA3_MTPWnoChild_genename.gmt       Polycistronic mtRNA processing   \n",
       "57  MITOCARTA3_MTPWnoChild_genename.gmt              Phospholipid metabolism   \n",
       "0   MITOCARTA3_MTPWnoChild_genename.gmt           Amidoxime reducing complex   \n",
       "64  MITOCARTA3_MTPWnoChild_genename.gmt                                  SAM   \n",
       "53  MITOCARTA3_MTPWnoChild_genename.gmt                                  OXA   \n",
       "52  MITOCARTA3_MTPWnoChild_genename.gmt  Nucleotide synthesis and processing   \n",
       "51  MITOCARTA3_MTPWnoChild_genename.gmt                    Nucleotide import   \n",
       "50  MITOCARTA3_MTPWnoChild_genename.gmt                            Mitophagy   \n",
       "49  MITOCARTA3_MTPWnoChild_genename.gmt      Mitochondrial ribosome assembly   \n",
       "\n",
       "   Overlap       P-value  Adjusted P-value  Odds Ratio  Combined Score  \\\n",
       "55   46/60  3.628878e-11      3.193413e-09    2.331566       56.049712   \n",
       "48   53/76  4.433454e-10      1.950720e-08    2.137311       46.030574   \n",
       "5    26/34  1.235894e-06      3.625288e-05    2.227091       30.296708   \n",
       "10   12/12  4.458087e-06      9.807792e-05    2.831193       34.882532   \n",
       "54   24/49  4.105660e-02      6.951296e-01    1.395549        4.455712   \n",
       "63     3/3  4.739520e-02      6.951296e-01    2.772647        8.454449   \n",
       "8      6/9  6.230448e-02      7.832563e-01    1.902769        5.281557   \n",
       "62    8/23  6.378397e-01      9.968845e-01    0.994844        0.447350   \n",
       "61     1/5  8.956430e-01      9.968845e-01    0.748918        0.082541   \n",
       "60    3/12  8.708325e-01      9.968845e-01    0.767389        0.106134   \n",
       "59    4/25  9.937830e-01      9.968845e-01    0.477194        0.002976   \n",
       "58    1/12  9.957247e-01      9.968845e-01    0.326561        0.001399   \n",
       "57    4/19  9.548283e-01      9.968845e-01    0.628959        0.029073   \n",
       "0      1/2  5.941539e-01      9.968845e-01    1.653968        0.861084   \n",
       "64     2/3  2.989931e-01      9.968845e-01    1.973451        2.382616   \n",
       "53     1/1  3.627075e-01      9.968845e-01    2.760141        2.799220   \n",
       "52    7/17  4.239890e-01      9.968845e-01    1.183784        1.015743   \n",
       "51     1/7  9.579569e-01      9.968845e-01    0.547795        0.023529   \n",
       "50    4/11  6.102665e-01      9.968845e-01    1.077656        0.532211   \n",
       "49    4/19  9.548283e-01      9.968845e-01    0.628959        0.029073   \n",
       "\n",
       "                                                Genes  \n",
       "55  Ndufb10;Ndufb7;Ndufa13;Cox6c;Sdhb;Ndufa6;Ndufs...  \n",
       "48  Mrpl34;Mrpl21;Mrps6;Mrpl13;Mrps18b;Gadd45gip1;...  \n",
       "5   Ndufb10;Ndufb7;Ndufa13;Ndufa6;Ndufs5;Ndufb11;N...  \n",
       "10  Cox7b;Cox8a;Cox5a;Cox7a2;Cox6c;Cox4i1;Cox6b1;C...  \n",
       "54  Tmem70;Uqcc2;Ndufaf4;Cmc1;Timmdc1;Coa4;Tmem126...  \n",
       "63                              Higd2a;Higd1a;Cox7a2l  \n",
       "8              Uqcr11;Cyc1;Uqcrfs1;Uqcrh;Uqcrq;Uqcr10  \n",
       "62        Hagh;Prdx4;Gpx4;Prdx5;Sod1;Prdx3;Gpx1;Msrb2  \n",
       "61                                               Etfa  \n",
       "60                                     Mpc1;Pdhb;Mpc2  \n",
       "59                               Clpp;Lap3;Htra2;Parl  \n",
       "58                                           Hsd17b10  \n",
       "57                        Triap1;Chpt1;Stard7;Prelid1  \n",
       "0                                               Cyb5b  \n",
       "64                                          Mtx1;Mtx2  \n",
       "53                                              Oxa1l  \n",
       "52           Nt5dc2;Guk1;Nudt2;Dut;Suclg1;Hint1;Dguok  \n",
       "51                                            Slc25a5  \n",
       "50                            Park7;Parl;Tomm7;Fundc1  \n",
       "49                          Mtg1;Malsu1;Tfb1m;Mpv17l2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# EnrichR\n",
    "results = {}\n",
    "for group in genelists:\n",
    "    gene_symbols = pd.merge(genelists[group], conv, on='gene_id')[['external_gene_name']]\n",
    "\n",
    "    enr = gp.enrichr(   gene_list=gene_symbols,\n",
    "                        # gene_sets='/Users/daehwa/Library/CloudStorage/OneDrive-Personal/Junlab/Data_Library/MITOCARTA3.0/MITOCARTA3_MTPW_genename.gmt',\n",
    "                        gene_sets='/Users/daehwa/Library/CloudStorage/OneDrive-Personal/Junlab/Data_Library/MITOCARTA3.0/MITOCARTA3_MTPWnoChild_genename.gmt',\n",
    "                        # gene_sets='/Users/daehwa/Library/CloudStorage/OneDrive-Personal/Junlab/Data_Library/GSEA/gmt/m5.go.v2023.1.Mm.symbols.gmt',\n",
    "                        # gene_sets='/Users/daehwa/Library/CloudStorage/OneDrive-Personal/Junlab/Data_Library/GSEA/gmt/m5.go.cc.v2023.1.Mm.symbols.gmt',\n",
    "                        background=BG_Symbols,\n",
    "                        outdir=None, \n",
    "                        verbose=False\n",
    "                    )\n",
    "    tmp = enr.results.sort_values('Adjusted P-value')\n",
    "    tmp.to_csv(f'{group}_Enrich.tsv', sep='\\t', index=False)\n",
    "\n",
    "    results[group] = tmp\n",
    "\n",
    "    print(group)\n",
    "    display(tmp[:20])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "univ",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
