{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Seqname</th>\n",
       "      <th>CCDS_tag</th>\n",
       "      <th>Transcript_range</th>\n",
       "      <th>Transcript_length</th>\n",
       "      <th>5UTR_length</th>\n",
       "      <th>CDS_length</th>\n",
       "      <th>3UTR_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001.5</td>\n",
       "      <td>ENSMUST00000000001.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>108014596,108053462</td>\n",
       "      <td>3262</td>\n",
       "      <td>141</td>\n",
       "      <td>1065</td>\n",
       "      <td>2056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003.16</td>\n",
       "      <td>ENSMUST00000000003.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Yes</td>\n",
       "      <td>76881507,76897229</td>\n",
       "      <td>902</td>\n",
       "      <td>140</td>\n",
       "      <td>525</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028.16</td>\n",
       "      <td>ENSMUST00000000028.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>-</td>\n",
       "      <td>chr16</td>\n",
       "      <td>Yes</td>\n",
       "      <td>18599197,18630722</td>\n",
       "      <td>2143</td>\n",
       "      <td>313</td>\n",
       "      <td>1701</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000037.18</td>\n",
       "      <td>ENSMUST00000019101.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>+</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Yes</td>\n",
       "      <td>159945768,160041192</td>\n",
       "      <td>4842</td>\n",
       "      <td>501</td>\n",
       "      <td>2718</td>\n",
       "      <td>1623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000000049.12</td>\n",
       "      <td>ENSMUST00000000049.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>Yes</td>\n",
       "      <td>108286119,108305222</td>\n",
       "      <td>1190</td>\n",
       "      <td>51</td>\n",
       "      <td>1038</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52349</th>\n",
       "      <td>ENSMUSG00000118670.1</td>\n",
       "      <td>ENSMUST00000239545.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Muc19</td>\n",
       "      <td>+</td>\n",
       "      <td>chr15</td>\n",
       "      <td>No</td>\n",
       "      <td>91722531,91832447</td>\n",
       "      <td>22800</td>\n",
       "      <td>41</td>\n",
       "      <td>22575</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52350</th>\n",
       "      <td>ENSMUSG00000118671.1</td>\n",
       "      <td>ENSMUST00000239552.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Eppk1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr15</td>\n",
       "      <td>No</td>\n",
       "      <td>75973330,76004395</td>\n",
       "      <td>12332</td>\n",
       "      <td>1133</td>\n",
       "      <td>10377</td>\n",
       "      <td>822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52351</th>\n",
       "      <td>ENSMUSG00000118672.1</td>\n",
       "      <td>ENSMUST00000239554.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Muc4</td>\n",
       "      <td>+</td>\n",
       "      <td>chr16</td>\n",
       "      <td>No</td>\n",
       "      <td>32555015,32602596</td>\n",
       "      <td>12361</td>\n",
       "      <td>183</td>\n",
       "      <td>11973</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53206</th>\n",
       "      <td>ENSMUSG00001074846.1</td>\n",
       "      <td>ENSMUST00010126032.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Iqcf3</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9</td>\n",
       "      <td>Yes</td>\n",
       "      <td>106420585,106438830</td>\n",
       "      <td>1131</td>\n",
       "      <td>59</td>\n",
       "      <td>612</td>\n",
       "      <td>460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54449</th>\n",
       "      <td>ENSMUSG00002076083.1</td>\n",
       "      <td>ENSMUST00000105569.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Klhl17</td>\n",
       "      <td>-</td>\n",
       "      <td>chr4</td>\n",
       "      <td>Yes</td>\n",
       "      <td>156313792,156319314</td>\n",
       "      <td>2723</td>\n",
       "      <td>101</td>\n",
       "      <td>1923</td>\n",
       "      <td>699</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22410 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_id          transcript_id       Gene_type gene_name  \\\n",
       "0       ENSMUSG00000000001.5   ENSMUST00000000001.5  protein_coding     Gnai3   \n",
       "1      ENSMUSG00000000003.16  ENSMUST00000000003.14  protein_coding      Pbsn   \n",
       "2      ENSMUSG00000000028.16  ENSMUST00000000028.14  protein_coding     Cdc45   \n",
       "4      ENSMUSG00000000037.18  ENSMUST00000019101.11  protein_coding     Scml2   \n",
       "5      ENSMUSG00000000049.12   ENSMUST00000000049.6  protein_coding      Apoh   \n",
       "...                      ...                    ...             ...       ...   \n",
       "52349   ENSMUSG00000118670.1   ENSMUST00000239545.1  protein_coding     Muc19   \n",
       "52350   ENSMUSG00000118671.1   ENSMUST00000239552.1  protein_coding     Eppk1   \n",
       "52351   ENSMUSG00000118672.1   ENSMUST00000239554.1  protein_coding      Muc4   \n",
       "53206   ENSMUSG00001074846.1   ENSMUST00010126032.3  protein_coding     Iqcf3   \n",
       "54449   ENSMUSG00002076083.1   ENSMUST00000105569.5  protein_coding    Klhl17   \n",
       "\n",
       "      Strand Seqname CCDS_tag     Transcript_range  Transcript_length  \\\n",
       "0          -    chr3      Yes  108014596,108053462               3262   \n",
       "1          -    chrX      Yes    76881507,76897229                902   \n",
       "2          -   chr16      Yes    18599197,18630722               2143   \n",
       "4          +    chrX      Yes  159945768,160041192               4842   \n",
       "5          +   chr11      Yes  108286119,108305222               1190   \n",
       "...      ...     ...      ...                  ...                ...   \n",
       "52349      +   chr15       No    91722531,91832447              22800   \n",
       "52350      -   chr15       No    75973330,76004395              12332   \n",
       "52351      +   chr16       No    32555015,32602596              12361   \n",
       "53206      -    chr9      Yes  106420585,106438830               1131   \n",
       "54449      -    chr4      Yes  156313792,156319314               2723   \n",
       "\n",
       "       5UTR_length  CDS_length  3UTR_length  \n",
       "0              141        1065         2056  \n",
       "1              140         525          237  \n",
       "2              313        1701          129  \n",
       "4              501        2718         1623  \n",
       "5               51        1038          101  \n",
       "...            ...         ...          ...  \n",
       "52349           41       22575          184  \n",
       "52350         1133       10377          822  \n",
       "52351          183       11973          205  \n",
       "53206           59         612          460  \n",
       "54449          101        1923          699  \n",
       "\n",
       "[22410 rows x 12 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "P_gtf = pd.read_csv('/Data_2/Daehwa/Data_Library/GTF_parsed/v0.7.1/gencode.vM27.annotation.gtf/Processed_gtf.tsv', sep='\\t')\n",
    "P_gtf = P_gtf[['Transcript_ID', 'Gene_ID', 'Gene_type', 'Gene_name', 'Strand', 'Seqname', 'CCDS_tag', 'Transcript_range', 'Transcript_length', '5UTR_length', 'CDS_length', '3UTR_length']]\n",
    "P_gtf = P_gtf.rename(columns={'Transcript_ID':'transcript_id', 'Gene_ID':'gene_id', 'Gene_name':'gene_name'})\n",
    "\n",
    "repre = pd.read_csv('/Data_2/Daehwa/Data_Library/representative-isoforms.txt', sep='\\t', header=None, usecols=[0,1], names=['gene_id','transcript_id'])\n",
    "repre_GTF = pd.merge(repre, P_gtf, on=['gene_id','transcript_id'])\n",
    "\n",
    "# protein coding\n",
    "repre_GTF = repre_GTF[ (repre_GTF['5UTR_length']+repre_GTF['CDS_length']+repre_GTF['3UTR_length'])!=0 ]\n",
    "\n",
    "display(repre_GTF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>Tef</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <td>ENSMUSG00000022389.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <td>ENSMUST00000109553.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_type</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strand</th>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seqname</th>\n",
       "      <td>chr15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDS_tag</th>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Transcript_range</th>\n",
       "      <td>81686622,81711064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Transcript_length</th>\n",
       "      <td>4403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5UTR_length</th>\n",
       "      <td>363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDS_length</th>\n",
       "      <td>858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3UTR_length</th>\n",
       "      <td>3182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name                            Tef\n",
       "gene_id            ENSMUSG00000022389.16\n",
       "transcript_id      ENSMUST00000109553.10\n",
       "Gene_type                 protein_coding\n",
       "Strand                                 +\n",
       "Seqname                            chr15\n",
       "CCDS_tag                             Yes\n",
       "Transcript_range       81686622,81711064\n",
       "Transcript_length                   4403\n",
       "5UTR_length                          363\n",
       "CDS_length                           858\n",
       "3UTR_length                         3182"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Target genes\n",
    "genes = {'gene_name':['Tef']}\n",
    "genes = pd.DataFrame(genes)\n",
    "\n",
    "genes = pd.merge(genes, repre_GTF, on='gene_name').set_index('gene_name').T\n",
    "display(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Tef': 'chr15:77602291-85796617'}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Select range for genes\n",
    "margin = 0.05\n",
    "\n",
    "ranges = {}\n",
    "for gene in genes:\n",
    "\n",
    "    start = round(int(genes[gene]['Transcript_range'].split(',')[0]) * (1-margin))\n",
    "    end   = round(int(genes[gene]['Transcript_range'].split(',')[1]) * (1+margin))\n",
    "\n",
    "    ranges[gene] = f\"{genes[gene]['Seqname']}:{start}-{end}\"\n",
    "\n",
    "display(ranges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Counting coverage of reads\n",
    "# Here I used 'samtools depth'. If you want to calculate the coverage of paired sequencing bam files, use 'samtools mpileup' to prevent overlapped count.\n",
    "SPs = [f'D{day}{rep}' for day in ['0','4','8'] for rep in ['a','b','c']]\n",
    "\n",
    "for gene in ranges:\n",
    "    for SP in SPs:\n",
    "        !samtools depth -r {ranges[gene]} /Data_2/Jun/Adipocytes/alignments/novaseq/{SP}.sorted.bam | gzip > {gene}_{SP}_RPF_depth.tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Tef': 'ENSMUST00000109553.10:1-4403'}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Select range for genes\n",
    "\n",
    "ranges = {}\n",
    "for gene in genes:\n",
    "    \n",
    "    ranges[gene] = f\"{genes[gene]['transcript_id']}:1-{int(genes[gene]['Transcript_length'])}\"\n",
    "\n",
    "display(ranges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Counting coverage of reads\n",
    "# Here I used 'samtools depth'. If you want to calculate the coverage of paired sequencing bam files, use 'samtools mpileup' to prevent overlapped count.\n",
    "SPs = [f'D{day}{rep}' for day in ['0','4','8'] for rep in ['a','b','c']]\n",
    "\n",
    "for gene in ranges:\n",
    "    for SP in SPs:\n",
    "        !samtools depth -r {ranges[gene]} /Data_2/Jun/Adipocytes/tr-aln/novaseq/{SP}.rep.bam | gzip > {gene}_{SP}_RPF_T_depth.tsv.gz"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
